cEP-0021.md

```diff
@@ -0,0 +1,266 @@
+
+Bear writer can also invoke the Profiler by using `coafile` i.e.
+`profile_bears = XXX`.
+
+- The profiling will start by profiling the `run()` method of the bears
+  The advantage of profiling just the `run()` method is that, bear writer
+  doesn't have to spend time on loading the files, collecting the settings,
+  etc. as this is all done by coala itself, they just have to Profile
+  the `run()` method to create a Better Performant bear.
+
+After invoking the profiler on the run method of a bear it will first save
+all the profile output on a local temp file which be further used by
+`pstats module` to sort the profile results and formats them in a way desired
+by the Bear writer.
+
+> Currently default format is to strip the directory (`strip_dirs()`) and sort
+> the stats of profiler result by cummulative time (`cumtime`) and display the
+> profiler result on *CLI*.
+>
+> ```python
+> pstats.strip_dirs().sort_stats('cumtime')
+> ```
+
+Bear writer will have an option to dump the formatted result in a
+specified file, instead of displaying it on terminal. i.e. by using
+`filename/file-path` instead of `True`.
+
+```
+coala -b Bear Mybear --files <file> --profile-bears "<True or filename/file-path>"
+```
+
+Bear writer will also have an option to directly dump the raw profile output,
+which can be further used for estimation of profiler stats.
+
+```
+coala -b Bear Mybear --files <file> --profile-bears "True,dump"
+```
+
+# Implementation Details
+
+When user invokes the profiler with either `CLI` or by `coafile`
+then `execute` method of Bear Base Class will call the `run_bear_from_section`
+method.
+
+```python
+def run_bear_from_section(self, args, kwargs):
+    ............
+    profile_bears_results, args, kwargs = self.setup_profiler(
+        *args, **kwargs)
+    return profile_bears_results if profile_bears_results else self.run(
+        *args, **kwargs)
+```
+
+Which futher calls the `setup_profiler()` method, and `setup_profiler()` method
+first enables an instance of the `Profile` class, and then enable the profiler
+on `run()` method of bears then check whether keyword arguments
+passed to the Bear contain the `profile_bears` key or not.
+
+```python
+if profile_bears is None:
+    kwargs.pop('profile_bears')
+    return False, args, kwargs
+```
+
+- If `yes` then it will split all the values passed to
+  `--profile-bears` through CLI <br>i.e.
+  `--profile-bears "True,strip_dirs(),sort_stats('cumtime')"`
+  or by `coafile` and return a list of all the comma seperated values.
+  i.e. `['True','strip_dirs','sort_stats('cumtime')']`
+
+- If `No` then it will return `False` and no profiling will be done
+  and coala will continue it's normal execution.
+
+  > In both the case i.e. whether Profiler is getting invoked or not
+  > coala will continue it's normal execution.
+
+After that `setup_profiler` method check whether Bear writer directly want
+to `dump` the profile output <br>i.e. `--profile-bears "True,dump"`
```

Comment:
ah, now you're explaining the setting more, by accident :stuck_out_tongue: 

two thoughts:

- definitely try to clearly explain the setting and it's possible values further above where you introduce it.
- I don't really like that you want to have a setting that's basically a number of different settings in a format string. Is anything else at coala using a setting like that right now? While it's true that settings like the bear to be run are 'lists' and represented as comma separated values, your settings seem to be of very different nature. Maybe it's better to split this into two or three different settings? Also maybe you can re-use existing settings for things like where to pipe output.

Reply:
Actually reason of passing it as a string was that through cli we can't pass parenthesis so one way was to solve that problem was to parse it in string and create a list of different settings

Reply:
So I created a same method which will take care of both `CLI` and `coafile`

Reply:
Also I am not hindering the current settings for example of `bears` i.e. <br>  `-b PEP8Bear,PyUnusedCodeBear..` , I am creating a `split function` which will split only the settings passed to `--profile-bears` i.e. `--profile-bears "True,sort_stats('ncalls'), ` 

I have to make that thing because we can't pass parenthesis in `CLI` , so passing it like a string and after that listing all the settings, why I have created a custom function because suppose user passed something like `sort_stats(-1) ` so `-1` is a argument to make it work I have to extract the arg from the string and pass it to `sort_stats()` function to make it work.